File: Cpp/src/bitboard.cpp
// bitboard.cpp

#include "datatypes.hpp"
#include "bitboard.hpp"
#include "Board.hpp"

#include <bit>
#include <bitset>
#include <cstdint>
#include <iostream>

/*
    Bit Manipulation
*/

// Clears the least significant set bit and returns its index (0 - 63)
uint8_t pop_ls1b(Bitboard& bb) {
    uint8_t index = __builtin_ctzll(bb); // Use gcc/clang intrinsic
    bb &= ~1ULL << index;
    return index;
}

// Counts the number of set bits in a bitboard
uint8_t count_bits(Bitboard bb) {
    return (uint8_t)std::popcount(bb);
}

/*
    Misc
*/

void print_bitboard(Bitboard board) {

    std::cout << "\n";

    for (int rank = RANK_8; rank <= RANK_1; ++rank) {
        std::cout << 8 - rank << "   ";
        for (int file = FILE_A; file <= FILE_H; ++file) {
            uint8_t sq = FR2SQ(file, rank);
            if (GET_BIT(board, sq)) {
                std::cout << "X";
            }
            else {
                std::cout << "-";
            }
        }
        std::cout << "\n";
    }

    std::cout << "    ABCDEFGH\n\n";
    std::cout << "Bits set: " << (int)count_bits(board) << "\n";
    std::cout << "\n";
}

File: Python/src/bitboard.py
# bitboard.py: Bitboard operation functions. No class as overhead is too high

from datatypes import Squares, Mirror64

type Bitboard = int

# Common bitboard operations
def get_bit(bitboard: int, square: Squares) -> int:
    """
    Returns the bit at the specified square.
    """
    return (bitboard >> square.value) & 1
    
def set_bit(bitboard: int, square: Squares) -> int:
    """
    Sets the bit at the specified square.
    """
    return bitboard | (1 << square.value)
    
def clear_bit(bitboard: int, square: Squares) -> int:
    """
    Clears the bit at the specified square.
    """
    return bitboard & ~(1 << square.value)

def pop_ls1b(bitboard: int) -> tuple[Squares, int]:
    """
    Clears the least significant set bit and returns its square index
    """
    bitboard &= 0xFFFFFFFFFFFFFFFF  # Ensure unsigned 64-bit
    if bitboard == 0:
        print("Warning: Attempted to pop from an empty bitboard.")
        return Squares.NO_SQ, bitboard
    
    index = (bitboard & -bitboard).bit_length() - 1
    bb = bitboard & (bitboard - 1)
    return Squares(index), bb
        
def count_bits(bitboard: int) -> int:
    """
    Counts the number of bits set in the bitboard.
    """
    return (bitboard & 0xFFFFFFFFFFFFFFFF).bit_count()

def print_bitboard(bitboard: int) -> None:
    """
    Prints the bitboard in a human-readable format.
    """
    print("\n")
    for rank in range(8):
        print(f"{8 - rank}   ", end="")
        for file in range(8):
            sq = rank * 8 + file
            if get_bit(bitboard, Squares(sq)):
                print("X", end="")
            else:
                print("-", end="")
        print()
    print("    ABCDEFGH\n")
    print(f"Bits set: {count_bits(bitboard)}")
    print(f"Value: {bitboard}\n")
    return